fontconfig/src/lib.rs

```diff
@@ -227,6 +227,31 @@ impl<'fc> Pattern<'fc> {
         }
     }
 
+    /// Get the languages set of this pattern.
+    pub fn get_lang_set<'a>(&'a self) -> Option<Vec<&'a str>> {
+        unsafe {
+            let mut ret: *mut sys::FcLangSet = ptr::null_mut();
+            if ffi_dispatch!(LIB, FcPatternGetLangSet, self.pat, FC_LANG.as_ptr(), 0, &mut ret as *mut _)
+                == sys::FcResultMatch
+            {
+                let ss: *mut sys::FcStrSet = ffi_dispatch!(LIB, FcLangSetGetLangs, ret);
+                let lang_strs: *mut sys::FcStrList = ffi_dispatch!(LIB, FcStrListCreate, ss);
+                let mut res = vec![];
+                loop {
+                    let lang_str: *mut sys::FcChar8 = ffi_dispatch!(LIB, FcStrListNext, lang_strs);
+                    if lang_str == ptr::null_mut() {
+                        break;
+                    }
+                    let cstr = CStr::from_ptr(lang_str as *const c_char);
+                    res.push(cstr.to_str().unwrap());
+                }
```

Comment:
The iterator created by `FcStrListCreate` is currently leaked. `FcStrListDone` needs to be called when done with the iterator. That can't be done in this method because it returns strings yielded by the iterator. I think the API needs to be reworked to be like [FontSet:](https://github.com/yeslogic/fontconfig-rs/blob/f23345ba41c8b3a2c1842eef4b74ef3d582bf437/fontconfig/src/lib.rs#L329)

- Create a `StrList` wrapper type to hold the the `FcStrList`
- Implement `from_raw`, `iter`, and `drop` methods on it
- In the `drop` impl call `FcStrListDone`

Reply:
Thanks for the pointers! I'll give it a try and update the PR.

Reply:
Maybe I'm missing the obvious, what would be wrong with copying each `lang_str` and storing it as an owned `String` rather than providing references with little guarantee of validity or lifetime?
Client code would need to perform a copy anyway, before eventually calling `FcStrListDone` through `LangSet.drop`.
Impl would look so:
```rust
pub fn get_lang_set<'a>(&'a self) -> Option<Vec<String>> {
        unsafe {
            let mut ret: *mut sys::FcLangSet = ptr::null_mut();
            if ffi_dispatch!(LIB, FcPatternGetLangSet, self.pat, FC_LANG.as_ptr(), 0, &mut ret as *mut _)
                == sys::FcResultMatch
            {
                let ss: *mut sys::FcStrSet = ffi_dispatch!(LIB, FcLangSetGetLangs, ret);
                let lang_strs: *mut sys::FcStrList = ffi_dispatch!(LIB, FcStrListCreate, ss);
                let mut res = vec![];
                loop {
                    let lang_str: *mut sys::FcChar8 = ffi_dispatch!(LIB, FcStrListNext, lang_strs);
                    if lang_str == ptr::null_mut() {
                        break;
                    }
                    let cstr = CStr::from_ptr(lang_str as *const c_char);
                    res.push(cstr.to_str().unwrap().clone().to_owned());
                }
                ffi_dispatch!(LIB, FcStrListDone, lang_strs);
                Some(res)
            } else {
                None
            }
        }
    }
```

Reply:
> what would be wrong with copying each `lang_str` and storing it as an owned `String` 

Nothing wrong per se, it's just some allocation and copying that can be avoided.

> rather than providing references with little guarantee of validity or lifetime?

With the iterator option the lifetime of the strings that the iterator outputs would be tied to the `StrList` wrapper so they would remain valid while the `StrList` remained alive.

> Client code would need to perform a copy anyway

Not necessarily, as in your original example you might only want to compare the strings to a known value (`        .filter(|p| p.get_lang_set().unwrap().contains(&"ja"))`)

Here's a rough sketch of what I'm imagining. If you don't think it's worth pursing then we can go with the vector of owned strings.

```rust
pub struct StrList<'a> {
    list: *mut FcStrList,
    _life: &'a PhantomData
}

impl Pattern {
    pub fn lang_set(&self) -> Option<StrList<'_>> {
        unsafe {
            let mut ret: *mut sys::FcLangSet = ptr::null_mut();
            if ffi_dispatch!(LIB, FcPatternGetLangSet, self.pat, FC_LANG.as_ptr(), 0, &mut ret as *mut _) == sys::FcResultMatch {
                let ss: *mut sys::FcStrSet = ffi_dispatch!(LIB, FcLangSetGetLangs, ret);
                let lang_strs: *mut sys::FcStrList = ffi_dispatch!(LIB, FcStrListCreate, ss);
                Some(StrList::from_raw(self.fc, lang_strs))
            } else {
                None
            }
        }
    }
}

impl<'a> StrList<'a> {
    unsafe fn from_raw(_: &Fontconfig, raw_list: *mut sys::FcStrSet) -> Self {
        Self { list: raw_list, _life: &PhantomData }
    }
}

impl<'a> Drop for StrList<'a> {
    fn drop(self) {
        unsafe { ffi_dispatch!(LIB, FcStrListDone, self.list) };
    }
}

impl<'a> Iterator<Output = &'a str> {
    fn next(&mut self) -> Option<&'a str> {
        let lang_str: *mut sys::FcChar8 = unsafe { ffi_dispatch!(LIB, FcStrListNext, self.list) };
        if lang_str.is_null() {
            None
        }
        else {
            Some(CStr::from_ptr(lang_str as *const c_char))
        }
    }
}
```

Reply:
Indeed, I realized only later than my own example doesn't require owned strings at all.

Thanks a lot for the constructive discussion, it's very much appreciated!
The PhantomData for the  lifetime is a nice trick to know as well.

Now there's one bit I'm not too sure about, I'll start a thread about it.